Clickhouse Sessions + Frontend Sessions updates

mariusandra opened the pull request:
## Changes

- Frontend
  - Convert Sessions to TypeScript
  - Add selected day to URL to make back/forward work
  - Add back/forward buttons to the date selection
  - Simplify date parameter sent to the backend (might have a timezone issue?)
  - Refactor the logic quite a bit
  - Fix bug with order and timestamps in pagination of events

- Backend
  - Add limit/offset support for clickhouse sessions query
  - Add support for fetching events inside the sessions query
  - Refactor how the Postgres sessions query does pagination
  - Add Elements into the clickhouse response (might do async in frontend? or join in query? not sure)
  - Add user properties to the clickhouse response

## Checklist

- [ ] All querysets/queries filter by Team (if this PR affects any querysets/queries)
- [ ] Backend tests (if this PR affects the backend)
- [ ] Cypress E2E tests (if this PR affects the front and/or backend)

mariusandra committed:
convert sessions table logic to TS

mariusandra committed:
convert rest of sessions to TS

mariusandra committed:
sessions table logic refactor, store date in the url

mariusandra committed:
add back/forward buttons

mariusandra committed:
load sessions based on the URL, not after mount --> avoids duplicate query if opening an url with a filter

mariusandra committed:
prevent multiple queries

mariusandra committed:
throw error if failed instead of returning an empty list

mariusandra committed:
date from filters

mariusandra committed:
rename offset to nextOffset

mariusandra committed:
initial limit/offset block

mariusandra committed:
indent sql

mariusandra committed:
support limit + offset

mariusandra committed:
load LIMIT+1 sessions in postgres, pop last and show load more sign. (was: show sign if exactly LIMIT fetched)

mariusandra committed:
based offset is always 0

mariusandra committed:
default limit to 50

mariusandra committed:
events in clickhouse sessions

mariusandra committed:
add elements to query results

mariusandra commented:
Elements are now in the query and almost everything works. There's still a bug with the "order" column if the events list is paginated... and I need to get user properties into [the query](https://github.com/PostHog/posthog/blob/cd9d2423130317deee655cdd6abef3e2c2e1fde9/ee/clickhouse/queries/clickhouse_sessions.py#L20) somehow.

@EDsCODE @fuziontech what's the best way to get user properties now? What tables/materialized_views/etc should I get this from?

![image](https://user-images.githubusercontent.com/53387/94584413-deadd080-027e-11eb-9d05-3155a28b7c76.png)

mariusandra committed:
add person properties to sessions query response

mariusandra committed:
show seconds with two digits

mariusandra committed:
fix pagination, timestamp calculation and ordering on pages 2 and beyond

mariusandra committed:
mypy

mariusandra committed:
fix test

mariusandra committed:
add default time to fix test, fix some any(*) filter issues

mariusandra committed:
remove reverse

mariusandra commented:
The last todo items are fixed now. Ready for a review!

EDsCODE commented on posthog/api/insight.py:
```diff
@@ -152,12 +153,14 @@ def session(self, request: request.Request, *args: Any, **kwargs: Any) -> Respon
-            if len(result["result"]) > 49:
+            offset = filter.offset + limit - 1
+            if len(result["result"]) > SESSIONS_LIST_DEFAULT_LIMIT:
+                result["result"].pop()
```
fun!

EDsCODE reviewed (commented):
looks good! 

would be good to add a proper test for pagination session list now

mariusandra commented:
Will do, though only tomorrow morning. Feel free to merge it in until then if you need the stuff in master already

EDsCODE committed:
merge master

timgl commented:
Also failing mypy

mariusandra committed:
fix mypy error, indent SQL

mariusandra committed:
add pagination test for postgres

mariusandra committed:
Run insight API tests in Clickhouse

mariusandra commented:
I added a test to test the session pagination... and I enabled almost all of the insight API tests in clickhouse. 

The ones failing were funnels and retention. So I commented them out for now (only when testing EE) and added a `TODO` to fix this later.

![image](https://user-images.githubusercontent.com/53387/94742916-ca4cff00-0376-11eb-88fb-1f3e6ff76e07.png)

However, locally, I can't run the API tests unless I comment out these lines:

![image](https://user-images.githubusercontent.com/53387/94742991-e8b2fa80-0376-11eb-9606-4e00729a0328.png)

The error `e` says `cannot import name 'ClickhouseActions' from 'ee.clickhouse.views'` and all tests that make an API call fail with the error `Endpoint not found`.

Any idea why this could be happening?

EDsCODE commented:
Doesthis happen when you specifically target a test to run like ‘ee.clickhouse.views.test....’ but not when you run ‘ee.clickhouse’?

There’s some sort of circular dependency happening I believe with the barrel file for views. @timgl fixed it in the work he’s doing.

mariusandra commented:
You're right. If I run all the tests, everything passes. I guess this is safe to merge...
